x-pack/plugins/apm/server/lib/services/get_service_dependencies/index.ts

```diff
@@ -43,187 +49,254 @@ export type ServiceDependencyItem = {
+            : null,
+        timeseries: mergedMetrics.timeseries.map((point) => ({
+          x: point.x,
+          y: point.count > 0 ? point.latency_sum / point.count : null,
+        })),
+      },
+      throughput: {
+        value:
+          mergedMetrics.value.count > 0
+            ? calculateThroughput({
+                start,
+                end,
+                value: mergedMetrics.value.count,
+              })
+            : null,
+        timeseries: mergedMetrics.timeseries.map((point) => ({
+          x: point.x,
+          y:
+            point.count > 0
+              ? calculateThroughput({ start, end, value: point.count })
+              : null,
+        })),
+      },
+      errorRate: {
+        value:
+          mergedMetrics.value.count > 0
+            ? (mergedMetrics.value.error_count ?? 0) / mergedMetrics.value.count
+            : null,
+        timeseries: mergedMetrics.timeseries.map((point) => ({
+          x: point.x,
+          y: point.count > 0 ? (point.error_count ?? 0) / point.count : null,
+        })),
+      },
+    };
+
+    if (item.service) {
+      return {
+        name: item.service.name,
+        type: 'service' as const,
+        serviceName: item.service.name,
+        environment: item.service.environment,
+        // agent.name should always be there, type returned from joinByKey is too pessimistic
+        agentName: item.agent!.name,
+        ...destMetrics,
+      };
+    }
+
+    return {
+      name: item.span.destination.service.resource,
+      type: 'external' as const,
+      spanType: item.span.type,
+      spanSubtype: item.span.subtype,
+      ...destMetrics,
+    };
+  });
+
+  const latencySums = metricsByResolvedAddress
+    .map(
+      (metric) => (metric.latency.value ?? 0) * (metric.throughput.value ?? 0)
+    )
+    .filter(isFiniteNumber);
+
+  const minLatencySum = Math.min(...latencySums);
+  const maxLatencySum = Math.max(...latencySums);
+
+  return metricsByResolvedAddress.map((metric) => {
+    const impact =
+      isFiniteNumber(metric.latency.value) &&
+      isFiniteNumber(metric.throughput.value)
+        ? ((metric.latency.value * metric.throughput.value - minLatencySum) /
+            (maxLatencySum - minLatencySum)) *
+          100
+        : 0;
+
+    return {
+      ...metric,
+      impact,
+    };
+  });
+}
+
 export function getServiceDependencies({
   setup,
   serviceName,
   environment,
   numBuckets,
+  comparisonStart,
+  comparisonEnd,
 }: {
   serviceName: string;
   setup: Setup & SetupTimeRange;
   environment?: string;
   numBuckets: number;
+  comparisonStart: number;
+  comparisonEnd: number;
 }): Promise<ServiceDependencyItem[]> {
   return withApmSpan('get_service_dependencies', async () => {
     const { start, end } = setup;
-    const [allMetrics, destinationMap] = await Promise.all([
+    const commonProps = {
+      setup,
+      serviceName,
+      environment,
+      numBuckets,
+    };
+    const [
```

Comment:
I'm thinking this would be more straightforward if you'd have something like:

```ts
const [ current, previous ] = await Promise.all([ getServiceDependencies({ start, end }), getServiceDependencies({ start: comparisonStart, end: comparisonEnd }) ]
```

you can probably do that at the route level even. And then remove items from `previous` that are not in current.

Also, we might need to look into making sure that destinations are not resolved to different addresses for different time ranges. This might happen due to the fact that we sample span events.

Reply:
Resolving by the destination you mean for example:

Current period:
Outgoing connections: `postgresql ` and `opbeans:3000`
Incoming connections: `opbeans-java` and `opbeand-node`
Current Period metrics: `postgresql` and `opbeand-java` (Randomly picks `opbeans-java` or `opbeand-node`)

Previous period:
Outgoing connections: `postgresql ` and `opbeans:3000`
Incoming connections: `opbeans-java` and `opbeand-node`
Current Period metrics: `postgresql` and `opbeand-java` (picks the same destination as used in the current period)

Is it what you mean?